### ISSUE TITLE
Deploying to heroku with specific lucky version causes error

### ISSUE THREAD

**Issue by edwardloveall** [luckyframework/lucky#244]
Here's the commit I'm working off of: https://github.com/edwardloveall/suggestotron-lucky/tree/311a6b9094afbf7fb174f7172ff72daf75ce3c9e

When I try to deploy to heroku, it triggers a compile error that the method `create` doesn't exist on my form. I'm using a [specific commit of lucky](https://github.com/edwardloveall/suggestotron-lucky/blob/311a6b9094afbf7fb174f7172ff72daf75ce3c9e/shard.yml#L18) to deploy, which may very well be the problem.

Heroku is a free dyno using the `heroku/nodejs` and https://github.com/luckyframework/heroku-buildpack-crystal buildpacks.

<details>
<summary>Full deploy text with error</summary>

```
Counting objects: 250, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (228/228), done.
Writing objects: 100% (250/250), 83.64 KiB | 0 bytes/s, done.
Total 250 (delta 104), reused 0 (delta 0)
Compressing source files... done.
Building source:

-----> Node.js app detected

-----> Creating runtime environment
       
       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_VERBOSE=false
       NODE_ENV=production
       NODE_MODULES_CACHE=true

-----> Installing binaries
       engines.node (package.json):  unspecified
       engines.npm (package.json):   unspecified (use default)
       engines.yarn (package.json):  unspecified (use default)
       
       Resolving node version 8.x...
       Downloading and installing node 8.9.1...
       Using default npm version: 5.5.1
       Resolving yarn version 1.x...
       Downloading and installing yarn (1.3.2)...
       Installed yarn 1.3.2

-----> Restoring cache
       Skipping cache restore (not-found)

-----> Building dependencies
       Installing node modules (yarn.lock)
       yarn install v1.3.2
       [1/4] Resolving packages...
       [2/4] Fetching packages...
       info fsevents@1.1.2: The platform "linux" is incompatible with this module.
       info "fsevents@1.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
       [3/4] Linking dependencies...
       warning " > postcss-cssnext@3.0.2" has unmet peer dependency "caniuse-lite@^1.0.30000697".
       [4/4] Building fresh packages...
       Done in 115.79s.
       Running heroku-postbuild (yarn)
       yarn run v1.3.2
       $ bin/compile_assets build --production
       $ /tmp/build_cb6a762436f164eec23c07132f230398/node_modules/.bin/brunch build --production
       02:08:05 - info: compiled 6 files into 2 files in 2.3 sec
       Done in 2.98s.

-----> Caching build
       Clearing previous node cache
       Saving 2 cacheDirectories (default):
       - node_modules
       - bower_components (nothing to cache)

-----> Build succeeded!
 !     Unmet dependencies don't fail yarn install but may cause runtime issues
       https://github.com/npm/npm/issues/7494

-----> Crystal app detected
-----> Installing Crystal (0.23.1 due to latest release at https://github.com/crystal-lang/crystal) from https://github.com/crystal-lang/crystal/releases/download/0.23.1/crystal-0.23.1-3-linux-x86_64.tar.gz to /tmp/build_cb6a762436f164eec23c07132f230398/.heroku/crystal... done
-----> Installing Dependencies... Updating https://github.com/luckyframework/web.git
Updating https://github.com/luckyframework/cli.git
Updating https://github.com/mosop/teeplate.git
Updating https://github.com/luckyframework/habitat.git
Updating https://github.com/luckyframework/inflector.git
Updating https://github.com/luckyframework/record.git
Updating https://github.com/kostya/blank.git
Updating https://github.com/crystal-lang/crystal-db.git
Updating https://github.com/will/crystal-pg.git
Updating https://github.com/luckyframework/router.git
Updating https://github.com/jwaldrip/shell-table.cr.git
Updating https://github.com/luckyframework/migrator.git
Installing lucky_web (dc0a7391ace6032b9041f1cc62524352157c98a7)
Installing lucky_cli (0.6.2)
Installing teeplate (0.4.5)
Installing habitat (0.2.0)
Installing lucky_inflector (0f2acf61b5a4b2b514f596bf1f1706b6e9d963d6)
Installing lucky_record (0.2.1)
Installing blank (9106c23080fb08f271dcd00d4de75fd6481a1de6)
Installing db (0.4.3)
Installing pg (0.13.4)
Installing lucky_router (0.2.0)
Installing shell-table (0.9.2)
Installing lucky_migrator (bd203f9fa8eb851cce33b322d46060afc896e096)
done
-----> Compiling src/suggestotron.cr (auto-detected from shard.yml)... Error in src/suggestotron.cr:1: while requiring "./server.cr"

require "./server.cr"
^

in src/server.cr:21: instantiating 'HTTP::Server#listen()'

server.listen
       ^~~~~~

in .heroku/crystal/src/http/server.cr:160: instantiating 'listen(Bool)'

  def listen(reuse_port = false)
  ^

in .heroku/crystal/src/http/server.cr:163: expanding macro

      spawn handle_client(server.accept?)
      ^

in macro 'spawn' /tmp/build_cb6a762436f164eec23c07132f230398/.heroku/crystal/src/concurrent.cr:95, line 9:

   1.   
   2.     ->(
   3.       
   4.         __arg0 : typeof(server.accept?),
   5.       
   6.       
   7.       ) {
   8.       spawn(name: nil) do
>  9.         handle_client(
  10.           
  11.             __arg0,
  12.           
  13.           
  14.         )
  15.       end
  16.     
  17.       }.call(server.accept?)
  18.     
  19.   

instantiating 'handle_client((TCPSocket | Nil))'
in .heroku/crystal/src/http/server.cr:190: instantiating 'HTTP::Server::RequestProcessor#process((OpenSSL::SSL::Socket::Server | TCPSocket), (OpenSSL::SSL::Socket::Server | TCPSocket))'

    @processor.process(io, io)
               ^~~~~~~

in .heroku/crystal/src/http/server/request_processor.cr:16: instantiating 'process((OpenSSL::SSL::Socket::Server | TCPSocket), (OpenSSL::SSL::Socket::Server | TCPSocket), IO::FileDescriptor)'

  def process(input, output, error = STDERR)
  ^

in .heroku/crystal/src/http/server/request_processor.cr:39: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

          @handler.call(context)
                   ^~~~

in .heroku/crystal/src/http/server/handlers/compress_handler.cr:12: expanding macro

    {% if flag?(:without_zlib) %}
    ^

in macro 'macro_150363872' /tmp/build_cb6a762436f164eec23c07132f230398/.heroku/crystal/src/http/server/handlers/compress_handler.cr:12, line 12:

   1. 
   2.       request_headers = context.request.headers
   3. 
   4.       if request_headers.includes_word?("Accept-Encoding", "gzip")
   5.         context.response.headers["Content-Encoding"] = "gzip"
   6.         context.response.output = Gzip::Writer.new(context.response.output, sync_close: true)
   7.       elsif request_headers.includes_word?("Accept-Encoding", "deflate")
   8.         context.response.headers["Content-Encoding"] = "deflate"
   9.         context.response.output = Flate::Writer.new(context.response.output, sync_close: true)
  10.       end
  11. 
> 12.       call_next(context)
  13.     

instantiating 'call_next(HTTP::Server::Context)'
in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in .heroku/crystal/src/http/server/handlers/error_handler.cr:15: instantiating 'call_next(HTTP::Server::Context)'

      call_next(context)
      ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in .heroku/crystal/src/http/server/handlers/log_handler.cr:12: instantiating 'call_next(HTTP::Server::Context)'

    call_next(context)
    ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/charms/static_file_handler.cr:10: instantiating 'super(HTTP::Server::Context)'

    super(context)
    ^~~~~

in .heroku/crystal/src/http/server/handlers/static_file_handler.cr:29: instantiating 'call_next(HTTP::Server::Context)'

        call_next(context)
        ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in .heroku/crystal/src/http/server/handlers/static_file_handler.cr:29: instantiating 'call_next(HTTP::Server::Context)'

        call_next(context)
        ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in .heroku/crystal/src/http/server/handlers/websocket_handler.cr:39: instantiating 'call_next(HTTP::Server::Context)'

      call_next(context)
      ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/lucky_web/error_handler.cr:14: instantiating 'call_next(HTTP::Server::Context)'

    call_next(context)
    ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/lucky_web/flash.cr:17: instantiating 'call_next(HTTP::Server::Context)'

        call_next(context)
        ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/lucky_web/http_method_override_handler.cr:11: instantiating 'call_next(HTTP::Server::Context)'

    call_next(context)
    ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/lucky_web/log_handler.cr:15: instantiating 'call_next(HTTP::Server::Context)'

    call_next(context)
    ^~~~~~~~~

in .heroku/crystal/src/http/server/handler.cr:24: instantiating '(HTTP::Handler | Proc(HTTP::Server::Context, Nil))#call(HTTP::Server::Context)'

      next_handler.call(context)
                   ^~~~

in lib/lucky_web/src/lucky_web/route_handler.cr:10: instantiating 'LuckyWeb::Action+#perform_action()'

      handler.payload.new(context, handler.params).perform_action
                                                   ^~~~~~~~~~~~~~

in lib/lucky_web/src/lucky_web/renderable.cr:32: instantiating 'call()'

    response = call
               ^~~~

in macro 'action' /tmp/build_cb6a762436f164eec23c07132f230398/lib/lucky_web/src/lucky_web/routeable.cr:36, line 3:

   1.     infer_route(singular: false)
   2. 
>  3.     setup_call_method( #<loc:push>begin #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,5>#<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,5>TopicForm.create(#<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,22>params) #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,29>do |#<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,33>form, #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",3,39>topic|
   4.   #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",4,7>if #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",4,10>topic
   5.     #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",5,9>redirect(#<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",5,18>to: #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",5,22>Topics::Index)
   6.   else
   7.     #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",7,9>render(#<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",7,16>NewPage, #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",7,25>topic_form: #<loc:"/tmp/build_cb6a762436f164eec23c07132f230398/src/actions/topics/create.cr",7,37>form)
   8.   end
   9. end end#<loc:pop> )
  10.   

expanding macro
in macro 'setup_call_method' /tmp/build_cb6a762436f164eec23c07132f230398/lib/lucky_web/src/lucky_web/routeable.cr:10, line 7:

   1.     def call : LuckyWeb::Response
   2.       callback_result = run_before_callbacks
   3. 
   4.       response = if callback_result.is_a?(LuckyWeb::Response)
   5.         callback_result
   6.       else
>  7.         TopicForm.create(params) do |form, topic|
   8.   if topic
   9.     redirect(to: Topics::Index)
  10.   else
  11.     render(NewPage, topic_form: form)
  12.   end
  13. end
  14.       end
  15. 
  16.       callback_result = run_after_callbacks
  17. 
  18.       if callback_result.is_a?(LuckyWeb::Response)
  19.         callback_result
  20.       else
  21.         response
  22.       end
  23.     end
  24.   

undefined method 'create' for TopicForm:Class

 !     Push rejected, failed to compile Crystal app.

 !     Push failed
Verifying deploy...

!	Push rejected to gentle-plateau-42454.
```
</details>

**Comment by paulcsmith** [comment]
@edwardloveall I generated a project with the latest Lucky CLI and was able to deploy to Heroku. Here are the steps I ran:

* `heroku buildpacks:add https://github.com/heroku/heroku-buildpack-nodejs`
* `heroku buildpacks:add https://github.com/luckyframework/heroku-buildpack-crystal`
* `heroku config:set LUCKY_ENV=production`
* run `lucky gen.secret`
* `heroku config:set SECRET_KEY_BASE=<THE_GENERATED_SECRET>`
* `heroku addons:create heroku-postgresql:<PLAN_NAME>` <PLAN_NAME> could be `hobby-dev`

Then `git push master`. That worked ok for me. Maybe something in the latest version of master fixed the issue? In the meantime I'll pull down your commit and see if I can deploy it.

**Comment by paulcsmith** [comment]
@edwardloveall I pulled that branch down and it wouldn't work on Heroku or locally. I checked the shard.lock and it was because you're using the latest version of lucky, but lucky master is still locked to LuckyRevord 0.2.1, which doesn't have all the latest changes. So I don't think this is a Heroku issue or an issue with Lucky itself.

You may be able to override the version by setting `branch: master` for lucky record in your dependencies. I'll close this for now though. Feel free to reopen if it looks like I was incorrect and the issue is somewhere else

**Comment by edwardloveall** [comment]
Yup, that was it. (I also didn't have the secret key there, so thanks for that tip) That seems so obvious in retrospect but it's odd because it would compile locally. I ended up having to manually edit the shard.lock file to specify the commit. When I added lucky_record at a specific commit it wouldn't update the shard.lock file.

Here are the relevant sections of my shard.yml and shard.lock for anyone interested:

**shard.yml**
```yaml
  lucky_record:
    github: luckyframework/record
    commit: e535072
```

**shard.lock**
```yaml
  lucky_record:
    github: luckyframework/record
    commit: e535072cc890952ba3dc308670215eedb2914c64
```
